subsys/bluetooth/host/gatt.c

```diff
@@ -1910,44 +1910,75 @@ static void gatt_indicate_rsp(struct bt_conn *conn, uint8_t err,
 	}
 }
 
-static int gatt_send(struct bt_conn *conn, struct net_buf *buf,
-		     bt_att_func_t func, void *params)
+static struct bt_att_req *gatt_req_alloc(bt_att_func_t func, void *params,
+					 bt_att_write_buf_t write_buf,
+					 uint8_t op,
+					 size_t len)
 {
+	struct bt_att_req *req;
+
+	/* Allocate new request */
+	req = bt_att_req_alloc(BT_ATT_TIMEOUT);
+	if (!req) {
+		return NULL;
+	}
+
+#if defined(CONFIG_BT_SMP)
+	req->att_op = op;
+	req->len = len;
+	req->write_buf = write_buf;
+#endif
+	req->func = func;
+	req->user_data = params;
+
+	return req;
+}
+
+#ifdef CONFIG_BT_GATT_CLIENT
+static int gatt_req_send(struct bt_conn *conn, bt_att_func_t func, void *params,
+			 bt_att_write_buf_t write_buf, uint8_t op, size_t len)
+
+{
+	struct bt_att_req *req;
+	struct net_buf *buf;
 	int err;
 
-	if (params) {
-		struct bt_att_req *req;
+	req = gatt_req_alloc(func, params, write_buf, op, len);
+	if (!req) {
+		return -ENOMEM;
+	}
 
-		/* Allocate new request */
-		req = bt_att_req_alloc(BT_ATT_TIMEOUT);
-		if (!req) {
-			return -ENOMEM;
-		}
+	buf = bt_att_create_pdu(conn, op, len);
+	if (!buf) {
+		bt_att_req_free(req);
+		return -ENOMEM;
+	}
 
-		req->buf = buf;
-		req->func = func;
-		req->user_data = params;
+	req->buf = buf;
 
-		err = bt_att_req_send(conn, req);
-		if (err) {
-			bt_att_req_free(req);
-		}
-	} else {
-		err = bt_att_send(conn, buf, NULL, NULL);
+	err = write_buf(buf, len, params);
```

Comment:
Btw, we could in theory just encode when the request is actually about to be sent, meaning we don't have to allocate a buffer instead att.c would allocate one just as it does when resending and call write_buf, actually perhaps we should use the encode instead of write_buf.

Reply:
I have changed the name to 'encode'. However the change to just encode once sent I think is best to leave until after the release.